### ISSUE TITLE
Avoid converting Tensors to Scalars for ops that have Tensor (including optional) argument overload

### ISSUE THREAD

**Issue by sujoysaraswati** [pytorch/pytorch#67428]
## 🚀 Feature
Avoid converting Tensors to Scalars for ops that have Tensor variance

## Motivation
PyTorch has many ops that have both Scalar and Tensor variations. Sometimes, the overload is selected that converts a Tensor to Scalar, which causes a Tensor value from a backend device to be fetched onto the host for getting the Scalar argument.

This causes a bubble in execution (if we create lazy execution graphs, then these cause the graph to break as well). It would be good to allow ops to work on Tensor arguments, so that the ops can continue to work on tensors that are already in the backend device.

## Pitch

As per the order of overloads, Tensor overloads come first, before Scalar overloads. However, in case of the optional arguments like Scalar? and Tensor? variations (example: clamp), the order is not determined similarly. The optional argument ordering should also be similarly handled (#67429) 

Along with that, PyTorch also has an internal list of ops which allow converting numbers to Tensors. If more ops are allowed to convert numbers to Tensors, then it would reduce the number of cases Tensor values need to be sent to host. (#67435)

Example:
>>> import torch
>>> a = torch.Tensor(2, 3)
>>> b = torch.Tensor(3)
>>> c = a.clamp(0, b[0] - 1)  <- Here, clamp favors the Scalar? overload and gets the b[0] to host as a scalar value. Instead, it should be possible to send both the arguments as Tensors so that the execution continues to be on device.

**Comment by jbschlosser** [comment]
Triage review to determine labels.

**Comment by sujoysaraswati** [comment]
> > The impact we see is that this breaks an execution graph on the device.
> 
> I see that it fixes this one problem, but that doesn't answer my question: what is the impact on all the other ops?

I think similar issues could be there in other ops which has both optional Scalar and optional Tensors as overloaded instances. It seems inconsistent that Tensor overloads are preferred over Scalar but no such ordering is imposed when the arguments are optional.

> > I see only a few ops are allowed to convert a number to a tensor. Is there a reason for this?
> 
> I don't know the full story behind this, cc @ngimel might have some ideas?
> 
> Also note that both of these changes and the corresponding PRs are BC-breaking and so would need to be handled with a lot of care. In particular, is this a hard breakage that we are happy with? Or do we want to go through a deprecation cycle?
> 
> Finally is there any rule on this in the array API standard? And it would be interesting to see how numpy solves this issue.

Thanks for connecting this to the larger discussion.

**Comment by sujoysaraswati** [comment]
> > Finally is there any rule on this in the array API standard?
> 
> Yes. I think this is the summary:
> 
> * All functions are only required to support _array_ inputs.
> * For all builtin Python operators, using Python scalars must be supported (e.g., `x + 1` must work). See https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars
> * 0-D arrays are arrays (not scalars), and the same dtype casting rules apply.
> 
> In practice, making `x + 1` work means that scalars must be supported in `x.__add__`. It is _not_ required that also `add(x, 1)` works, but in practice that is likely to be the easiest route to implement it (because typically `__add__` and `add` are implemented in one place).

Thanks for the detailed explanation @rgommers . If I understood correctly, the python rule for mixing scalars with arrays is to convert the scalar to the 0D array with same dtype. PyTorch seems to do similar ordering in the python parser for tensors and scalars, but it has this additional allow conversion flag which enables this conversion only for some ops. Additionally, the ordering is not imposed on optional arguments. If I use clamp(0, b[0 - 1) [b is a tensor], then doesn't the python API rules imply the scalar 0 be turned into a tensor in this case?

> > And it would be interesting to see how numpy solves this issue.
> 
> * NumPy is very flexible with coercing scalars (and other input types like lists) to arrays, so any function that accepts arrays also accepts those "array-likes".
> * NumPy has separate "array scalar" types to paper over some usability issues, which are not 0-D arrays but do share attributes with arrays. Think of them like a mix of Python scalars and arrays. They are considered a design mistake by now.
> 
> NumPy does not build an execution graph, so that making supporting Python scalars easier - for building a graph the scalar needs to be replaced by some custom object (and I think that's what JAX does, and other libraries as well).
Yes, this is the intention behind raising this issue as well. The scalar, especially if it is extracted from a device tensor, breaks the graph of device execution and adds synchronous DMA for getting the scalar to host. It would be good to know if this behavior can be avoided in favor of continuing the execution on device.

**Comment by rgommers** [comment]
> If I understood correctly, the python rule for mixing scalars with arrays is to convert the scalar to the 0D array with same dtype.

Yes indeed.

>  If I use `clamp(0, b[0 - 1)` [b is a tensor], then doesn't the python API rules imply the scalar 0 be turned into a tensor in this case?

That misses a bracket; I'll assume you meant `clamp(0, b[0] - 1)`. And no, I don't think it implies that. This is a composite statement, let's unpack it:
1. `b[0] - 1`: `b[0]` is still a tensor; the `-` operator must support scalars so `1` gets turned into a 0-D array with the same dtype as `b`
2. `clamp` is a function call, scalars don't have to be supported in function calls (the rules are about Python operators only), so it's okay to either do the conversion of `0` to a tensor or to raise an exception.

For (2), it'd be great for all operators to behave consistently in PyTorch, however the array API standard has no opinion here. 

> The scalar, especially if it is extracted from a device tensor, breaks the graph of device execution and adds synchronous DMA for getting the scalar to host. It would be good to know if this behavior can be avoided in favor of continuing the execution on device.

My sense is that "extracting a scalar from a tensor" happens very rarely. What cases are you thinking about here?

Things like `torch.sum(t)` and `t[0]` for a 1-D array `t` return 0-D tensors.

This is a bit different in NumPy, where for a 1-D array `x` the statement `x[0]` gives back an "array scalar" which is no longer an array (it's also not a Python scalar though, so an object like that could still live on-device in theory).

**Comment by sujoysaraswati** [comment]
> > If I understood correctly, the python rule for mixing scalars with arrays is to convert the scalar to the 0D array with same dtype.
> 
> Yes indeed.
> 
> > If I use `clamp(0, b[0 - 1)` [b is a tensor], then doesn't the python API rules imply the scalar 0 be turned into a tensor in this case?
> 
> That misses a bracket; I'll assume you meant `clamp(0, b[0] - 1)`.

Yes, my bad. I had meant clamp(0, b[0] - 1).

 And no, I don't think it implies that. This is a composite statement, let's unpack it:
> 
> 1. `b[0] - 1`: `b[0]` is still a tensor; the `-` operator must support scalars so `1` gets turned into a 0-D array with the same dtype as `b`
> 2. `clamp` is a function call, scalars don't have to be supported in function calls (the rules are about Python operators only), so it's okay to either do the conversion of `0` to a tensor or to raise an exception.
> 

I see, so the API rule applies to Python operators alone.

> For (2), it'd be great for all operators to behave consistently in PyTorch, however the array API standard has no opinion here.

Yes, agreed. It would be great if PyTorch functions also have a similar consistent behavior.
 
> > The scalar, especially if it is extracted from a device tensor, breaks the graph of device execution and adds synchronous DMA for getting the scalar to host. It would be good to know if this behavior can be avoided in favor of continuing the execution on device.
> 
> My sense is that "extracting a scalar from a tensor" happens very rarely. What cases are you thinking about here?

I have currently seen it with clamp, though this happens in a model where the clamp is in a loop.
 
> Things like `torch.sum(t)` and `t[0]` for a 1-D array `t` return 0-D tensors.
> 
> This is a bit different in NumPy, where for a 1-D array `x` the statement `x[0]` gives back an "array scalar" which is no longer an array (it's also not a Python scalar though, so an object like that could still live on-device in theory).

Essentially I was looking for some semantics that can keep the computation on device for longer.

**Comment by rgommers** [comment]
> I see, so the API rule applies to Python operators alone.

Indeed. The rationale was that expressions like `2*x + 1` are very common, and would be really annoying to write as `tensor(2)*x + tensor(1)`. However scalars as parameters to function calls where tensors are expected are much more rare, so there's no need to require implementations to be more flexible there (some libraries have a strong preference for type safety and avoiding the bit of overhead needed to check there).

> I have currently seen it with clamp, though this happens in a model where the clamp is in a loop.

I'm not sure I understand this sentence. `clamp` should always return a tensor?

**Comment by ngimel** [comment]
For `clamp`, it turn out that if both arguments are 0-D tensors, they are still turned to scalars in a synchronizing way, so apparently "extracting a scalar from a tensor" happens. Only if they are 1D tensors are they left as tensors:
```
In [20]: torch.cuda.set_sync_debug_mode("warn")

In [21]: a=torch.randn(5, device="cuda")

In [22]: out=torch.clamp(a, torch.zeros((), device="cuda"), torch.ones((), device="cuda"))
<ipython-input-22-a543919b44b2>:1: UserWarning: called a synchronizing CUDA operation (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:147.)
  out=torch.clamp(a, torch.zeros((), device="cuda"), torch.ones((), device="cuda"))

In [23]: out=torch.clamp(a, torch.zeros((), device="cuda").view(1), torch.ones((), device="cuda").view(1)) #no warning
```
Also, not that the kernel for the op itself as written in [23] will be slower than for the op that uses cpu scalars for upper and lower boundaries, because for [23] signature the kernel will be handling broadcasting, while for scalar signature it will be a contiguous kernel, but maybe this slowdown is better than synchronization penalty.

**Comment by sujoysaraswati** [comment]
> > I see, so the API rule applies to Python operators alone.
> 
> Indeed. The rationale was that expressions like `2*x + 1` are very common, and would be really annoying to write as `tensor(2)*x + tensor(1)`. However scalars as parameters to function calls where tensors are expected are much more rare, so there's no need to require implementations to be more flexible there (some libraries have a strong preference for type safety and avoiding the bit of overhead needed to check there).
> 
> > I have currently seen it with clamp, though this happens in a model where the clamp is in a loop.
> 
> I'm not sure I understand this sentence. `clamp` should always return a tensor?

@rgommers - the comment is about the arguments of clamp, not the return value. I checked the effect on XLA, here is what I find -

Test:
-----
import torch
import numpy as np
import torch_xla.core.xla_model as xm

device = xm.xla_device()
x = torch.randn(2,3,4).to(device)
y = torch.randn(1).to(device)
z = torch.randn(1).to(device)

c = x.clamp_(z[0], y[0]) #By default, clamp_ will convert z[0] and y[0] to scalar arguments
print(c.to('cpu'))
-----
Behavior with master branch:
--------------------------------
The clamp arguments z[0] and y[0] are converted to scalar and it breaks the graph for XLA. The XLA log shows three graphs for the test case  -
## BEGIN_GRAPH
IR {
  %0 = f32[1]{0} xla::device_data(), device=CPU:0
  %1 = f32[1]{0} aten::view(%0), output_size=(1)
  %2 = f32[] aten::view(%1), output_size=(), ROOT=0
}

## END_GRAPH

## BEGIN_GRAPH
IR {
  %0 = f32[1]{0} xla::device_data(), device=CPU:0
  %1 = f32[1]{0} aten::view(%0), output_size=(1)
  %2 = f32[] aten::view(%1), output_size=(), ROOT=0
}

## END_GRAPH

## BEGIN_GRAPH
IR {
  %0 = f32[] xla::device_data(), device=CPU:0
  %1 = f32[] xla::device_data(), device=CPU:0
  %2 = f32[2,3,4]{2,1,0} xla::device_data(), device=CPU:0
  %3 = f32[2,3,4]{2,1,0} aten::clamp(%2, %1, %0), ROOT=0
}

## END_GRAPH

The stack trace for these scalar item calls are the following -

     at::_local_scalar_dense
     ...
     at::Tensor::item
     ...
     torch::PythonArgs::scalarOptional
     torch::autograd::THPVariable_clamp_
     ...

Behavior with clamp to work with "Tensor?" (proposed change in this issue) 
---------------------------------------------------------------------------------
There is a single graph without any break and no scalar data sent to CPU for clamp argument -

## BEGIN_GRAPH
IR {
  %0 = f32[1]{0} xla::device_data(), device=CPU:0
  %1 = f32[1]{0} aten::view(%0), output_size=(1)
  %2 = f32[] aten::view(%1), output_size=()
  %3 = f32[1]{0} xla::device_data(), device=CPU:0
  %4 = f32[1]{0} aten::view(%3), output_size=(1)
  %5 = f32[] aten::view(%4), output_size=()
  %6 = f32[2,3,4]{2,1,0} xla::device_data(), device=CPU:0
  %7 = f32[2,3,4]{2,1,0} aten::max(%6, %5)
  %8 = f32[2,3,4]{2,1,0} aten::min(%7, %2), ROOT=0
}

## END_GRAPH

Do you recommend we allow clamp to prioritize "Tensor?" over "Scalar?" so that the execution in the device can continue? As @ngimel mentioned, these graph breaks are also followed by a synchronous operation for the CPU to get the scalar value.

**Comment by rgommers** [comment]
> Do you recommend we allow clamp to prioritize "Tensor?" over "Scalar?" so that the execution in the device can continue? As @ngimel mentioned, these graph breaks are also followed by a synchronous operation for the CPU to get the scalar value.

Yes, I agree with prioritizing `Tensor?`. The current behavior is pretty bad, I think it's fair to say it's a bug. 

> > > I have currently seen it with clamp, though this happens in a model where the clamp is in a loop.
> > 
> > 
> > I'm not sure I understand this sentence. `clamp` should always return a tensor?
> 
> @rgommers - the comment is about the arguments of clamp, not the return value.

Thanks for explaining. This was just a terminology thing. With _"My sense is that "extracting a scalar from a tensor" happens very rarely."_ I meant semantically, that the user gets their hands on an actual Python scalar. Rather than "this happens to be a thing the C++ code does under the hood".
